README.md
# rdwd
<img src="https://github.com/brry/rdwd/raw/master/misc/hex/hex.png" width="104" height="120">

`rdwd` is an [R](https://www.r-project.org/) package to select, download and read climate data from the 
German Weather Service (Deutscher Wetterdienst, DWD).  
The DWD provides thousands of datasets with weather observations online at 
[opendata.dwd.de](https://opendata.dwd.de/climate_environment/CDC/observations_germany/climate/).  
Since May 2019, `rdwd` also supports reading the Radolan (binary) raster data at 
[grids_germany](https://opendata.dwd.de/climate_environment/CDC/grids_germany/).

`rdwd` is available on CRAN:
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version-last-release/rdwd)](https://cran.r-project.org/package=rdwd) 
[![downloads](http://cranlogs.r-pkg.org/badges/rdwd)](https://www.r-pkg.org/services)
[![Rdoc](http://www.rdocumentation.org/badges/version/rdwd)](https://www.rdocumentation.org/packages/rdwd)
!["rdwd dependencies"](https://tinyverse.netlify.com/badge/rdwd)

It has been presented at [FOSDEM 2017](https://archive.fosdem.org/2017/schedule/event/geo_weather/)
and [UseR!2017](https://user2017.sched.com/event/Axr3/rdwd-manage-german-weather-observations) in Brussels and with a 5 Minute [video](https://youtu.be/KOYZPMMgiHo?t=233) at [e-Rum2020](https://milano-r.github.io/erum2020program/lightning-talks.html#rdwd-r-interface-to-german-weather-service-data),
featured in Rstudio's [data package list](https://rviews.rstudio.com/2017/02/17/january-new-data-packages), 
written about in [OSOR](https://joinup.ec.europa.eu/collection/open-source-observatory-osor/news/study-german-weather-data) and used e.g. for
[NDR: Starkregen im Norden](https://story.ndr.de/starkregen-im-norden/index.html).
Development of `rdwd` was triggered 2016 by flash flood research in Braunsbach 
([1](https://www.uni-potsdam.de/en/natriskchange/qualification-program/task-force-braunsbach-flash-flood-2016), [2](https://doi.org/10.1016/j.scitotenv.2018.02.241),
[3](https://doi.org/10.5675/HyWa_2017,3_1),
[4](https://publishup.uni-potsdam.de/frontdoor/index/index/docId/39488)).


```diff
- HELP NEEDED
- with the new 5-minute data (April 2022), the fileIndex etc are getting very big.
- ideas on package size reduction are welcome at  https://github.com/brry/rdwd/issues/35
```

### Documentation

A website with more information, examples, use cases and an interactive map of the DWD stations
can be found at <https://bookdown.org/brry/rdwd>


### Usage

Usage for observational weather data from the measuring stations usually looks something like the following:

```R
# Download and install (once only):
install.packages("rdwd")
# Load the package into library (needed in every R session):
library(rdwd)

# select a dataset (e.g. last year's daily climate data from Potsdam city):
link <- selectDWD("Potsdam", res="daily", var="kl", per="recent")

# Actually download that dataset, returning the local storage file name:
file <- dataDWD(link, read=FALSE)
# Read the file from the zip folder:
clim <- readDWD(file, varnames=TRUE) # can happen directly in dataDWD

# Inspect the data.frame:
str(clim)
# Quick time series graphic:
plotDWD(clim, "FM.Windgeschwindigkeit")
```

For data interpolated onto a 1 km raster, including radar data up to the last hour,
see the corresponding [chapter](https://bookdown.org/brry/rdwd/raster-data.html) on the website.


### App
Since April 2023, there is an [interactive app](https://brry.shinyapps.io/wetter/) to compare weather periods:

<img src="https://github.com/brry/rdwd/raw/master/misc/app.png" width="577" height="415">

With `rdwd::app()`, you can run this locally with cached data, i.e. faster responses.

### New to R

If you're new to R, these links might help you to get started:

- [install R & Rstudio](https://github.com/brry/course/#install)
- [brief introduction to R](https://github.com/brry/hour)
- [very large set of slides I use for my courses](https://github.com/brry/course/#slides)

back to `rdwd`:


### Installation

#### Normal
```R
install.packages("rdwd")
```

#### Latest version
```R
rdwd::updateRdwd()
# checks version and (if needed) calls  remotes::install_github("brry/rdwd", build_vignettes=TRUE)
```

#### Full
Suggested (not mandatory) dependencies:  
```R
install.packages("rdwd", dependencies="Suggests") 
```

- `RCurl` for indexFTP and selectDWD(..., current=TRUE)
- `data.table`, `bit64` for readDWD(..., fread=TRUE)
- `raster`, `R.utils`, `ncdf4`, `dwdradar` for readDWD with gridded data
- `readr` for readDWD.stand(..., fast=TRUE)
- `knitr`, `rmarkdown`, `testthat`, `roxygen2`, `devtools`, `remotes`, `XML`, `gsheet` for local testing, development and documentation
- `leaflet`, `OSMscale`, `sp` for interactive/static maps, see [OSMscale installation tips](https://github.com/brry/OSMscale#installation)
- `shiny` for the interactive weather comparison app

Note: on Linux (Ubuntu), install `RCurl` via the terminal (CTRL+ALT+T, note lowercase rcurl):
```
sudo apt install r-cran-rcurl
```


NEWS
-----------------------------
-----------------------------

Version 1.7.0 (2023-04-xx)

- `dataDWD`: dir default set to `locdir()`
- `findID`+`selectDWD`: new argument 'failempty' to stop instead of warn if name does not match
- `readDWD.grib2`: rgdal retirement warning added
- `dataDWD` + `readDWD`: new argument 'hr' to auto-merge historical and recent files

-----------------------------
-----------------------------

Version 1.6.0 (2022-05-30)

new functions

- `plotDWD`
- `locdir` (will replace fixed "DWDdata" folder in `dataDWD`)
- `readDWD.deriv`
- `readDWD.rklim`
- `readDWD.pdf`
- `validFileTypes` (charstring vector)

improvements

- `selectDWD`: is completely rewritten and now vectorizes expandingly
- reading functions: use latin1 encoding
- `dataDWD`: new argument *method*, new default for *overwrite*
- several improvements for input checks, messages, tests and documentation
- `metaInfo`: accepts custom index
- `readDWD.radar`: new argument *dividebyten* 
- `readDWD.grib2`: new argument *pack* 
- website: new use case daily radar files


-----------------------------
-----------------------------

Version 1.5.0 (2021-04-08)

API changes:

- readVars: new argument 'quiet' (for updateIndexes)
- readDWD.grib2: initial release
- projectRasterDWD: new argument adjust05 
- dataDWD: new default: dbin=TRUE

enhancements / bug fixes:

- readDWD.data: fread checks for system command unzip, prints final value in message
- dataDWD: unfound URL messages also link to website fileIndex in German locale
- updateRdwd: version and date comparison improved, print local version if newer than on github
- createIndex: improved recognition of "Beschreibung_Stationen" meta files
- dwdparams + updateIndexes: missing/duplicate Kurz entry messages improved
- readDWD: subfunction names in elegant message outside the loop to avoid interrupted progbars.
- indexFTP + dirDWD: trailing slashes removed with regexp
- fileType: more informative message for failed type determination 
- DEU: now derived from NUTS regions and with new CRS (along with EUR)
- runLocalTests: improved logfiles, false positive messages removed 

website changes:

- package structure moved to first chapter
- fread instructions expanded
- FTP Folder list expanded
- more messages / warnings suppressed that are irrelevant to readers
- linked to in seeAlso sections of suitable function documentations
- Website: redirected man pages are now correctly linked to (e.g. fileIndex -> index)
- Documentation and dataset links are now opened in a new window
- new use case: values at locations in grid
- FTP address updated everywhere


-----------------------------
-----------------------------

Version 1.4.0 (2020-07-31)

API changes:

- dataDWD: argument 'file' renamed to 'url' to avoid confusion with local file names
- readDWD: 'fread' default set to NA. This is an experiment, see issue #22 
- readDWD: now has argument type (determined by new function fileType) to replace the ever growing list of types
- dataDWD, readDWD, selectDWD: order of arguments changed to reflect importance
- plotRadar: gains arguments axes, las, zlim, col, mar, keeppar. main now defaults to (and correctly keeps) x@title.
- selectDWD: gains argument remove_dupli to ignore DWD file upload errors
- readDWD.*: All subfunctions message their identity and have a 'quiet' argument
- new function: rdwdquiet()
- readDWD.meta: station and bundesland names are now char instead of factor (since R 4.0.0)
- indexFTP: new argument 'fast' to read file tree with data.table
- createIndex: gains argument checklog
- runLocalTests: new arguments for devtools::check and checkIndex
- readVars: params now an explicit argument (defaults to dwdparams)

enhancements / bug fixes:

- EUR: extends further east for plots with large width compared to height
- updateRdwd: unloads package before installation and informs about the need to re-load
- checkIndex: new check for duplicate meta files, writes to logfile (append, no overwrite), nicer path print, logs modification time of index files
- lldist: fix error that occurred when a df with a single row was given
- runLocalTests: warning logfile cleared before writing, checkIndex results included, false positives removed, on github
- source code function argument sections all have line breaks (except readDWD.*)
- documentation source code now is written in markdown, source files have been reorganized
- cross-references in documentation improved
- metaInfo: from/to columns date conversion corrected
- indexFTP + dataDWD: up to date and centralized sleep information 
- dataDWD: Download failure message improved, refer to fileIndex section on homepage if applicable.
- readDWD.nc: ncdf warning suppressed
- selectDWD: combines all warnings within the loop to a single message

website changes:

- improved (and stand-alone) fileIndex section
- plotRadar used for nice maps
- huxtable fixed
- use global quiet option instead of explicit `quiet=TRUE` all the time
- history section added
- package schematic revamped completely


-----------------------------
-----------------------------

Version 1.3.1 (2020-02-18)

API changes:  
- projectRasterDWD: latlon=T/F replaced with targetproj
- readDWD.binary: output list element renamed to 'dat' (from 'data') for consistency with other functions
- dataDWD: new argument dbin
- new functions: updateRdwd, plotRadar

enhancements / bug fixes:  
- readDWD.data: now correctly reads 10 minute timestamps
- readDWD.meta: column widths identified more elegantly and safely
- indexFTP: warn about requests to https instead of ftp servers, new argument `exclude.latest.bin`
- addBorders: added sp to 'Suggests', moved DEU + EUR to inst/extdata to handle CRAN build+check on systems without sp installed

website changes:  
- list of FTP Folders now with clickable URLs
- RQ example added
- order of radar graphs changed for better comparison in pdf


-----------------------------
-----------------------------

Version 1.2.0 (2019-10-26)
Highlights:

R/dataDWD.R
# DWD Daten mit R runterladen, Wetter und Klimadaten in R
# Deutscher Wetterdienst R Daten download Klimastationen
# Weather Data Germany download with R, Climate Data Germany
#
#' @title Download data from the DWD CDC FTP Server
#' @description Get climate data from the German Weather Service (DWD) FTP-server.
#' The desired dataset is downloaded into `dir`.
#' If `read=TRUE`, it is also read and processed.\cr
#' `dataDWD` handles vectors of URLs,
#' displays progress bars (if the package `pbapply` is available)
#' and by default does not re-download data already in `dir`
#' (but see argument `force` to update files).\cr
#' To solve "errors in download.file: cannot open URL", see
#' <https://bookdown.org/brry/rdwd/fileindex.html>.\cr
#' 
#' @return Presuming downloading and processing were successful:
#'         if `read=TRUE`, the desired dataset
#'         (as returned by [readDWD()]),
#'         otherwise the filename as saved on disc
#'         (may have "_n" appended in name, see [newFilename()]).\cr
#'         If length(file)>1, the output is a list of outputs / vector of filenames.\cr
#'         The output is always invisible.
#' @author Berry Boessenkool, \email{berry-b@@gmx.de}, Jun-Oct 2016
#' @seealso [selectDWD()]. [readDWD()], [download.file()].\cr
#'          <https://bookdown.org/brry/rdwd>\cr
#'          Helpful for plotting: [berryFunctions::monthAxis()],
#'          see also [berryFunctions::climateGraph()]
#' @keywords data file
#' @importFrom utils tail download.file browseURL
#' @importFrom berryFunctions newFilename owa tmessage twarning tstop truncMessage
#' @importFrom pbapply pblapply
#' @importFrom stats runif
#' @export
#' @examples
#' \dontrun{ ## requires internet connection
#' # find FTP files for a given station name and file path:
#' link <- selectDWD("Fuerstenzell", res="hourly", var="wind", per="recent")
#' # download file:
#' fname <- dataDWD(link, dir=locdir(), read=FALSE) ; fname
#' # dir="DWDdata" is the default directory to store files
#' # unless force=TRUE, already obtained files will not be downloaded again
#' 
#' # read and plot file:
#' wind <- readDWD(fname, varnames=TRUE) ; head(wind)
#' metafiles <- readMeta(fname)          ; str(metafiles, max.level=1)
#' column_names <- readVars(fname)       ; head(column_names)
#' 
#' plot(wind$MESS_DATUM, wind$F, main="DWD hourly wind Fuerstenzell", col="blue",
#'      xaxt="n", las=1, type="l", xlab="Date", ylab="Hourly Wind speed  [m/s]")
#' berryFunctions::monthAxis(1)
#' 
#' 
#' # current and historical files, keep historical in the overlap time period:
#' link <- selectDWD("Potsdam", res="daily", var="kl", per="hr"); link
#' potsdam <- dataDWD(link, dir=locdir(), hr=4)
#' plot(TMK~MESS_DATUM, data=tail(potsdam,1500), type="l")
#' 
#' 
#' # With many files (>>50), use sleep to avoid getting kicked off the FTP server
#' #links <- selectDWD(res="daily", var="solar")
#' #sol <- dataDWD(links, sleep=20) # random waiting time after download (0 to 20 secs)
#' 
#' # Real life examples can be found in the use cases section of the vignette:
#' # browseURL("https://bookdown.org/brry/rdwd")
#' }
#' 
#' @param url    Char (vector): complete file URL(s) (including base and filename.zip) 
#'               as returned by [selectDWD()]. Can be a vector with several FTP URLs.
#' @param base   Single char: base URL that will be removed from output file names.
#'               DEFAULT: [`dwdbase`]
#' @param joinbf Logical: paste `base` and `file url` together?
#'               Needed mostly for data at [`gridbase`].
#'               DEFAULT: FALSE (selectDWD returns complete URLs already)
#' @param dir    Char: Writeable directory name where to save the downloaded file.
#'               Created if not existent. DEFAULT: [locdir()]
#' @param force  Logical (vector): always download, even if the file already exists in `dir`?
#'               Use NA to force re-downloading files older than 24 hours.
#'               Use a numerical value to force after that amount of hours.
#'               Use something like `c(Inf, 24)` or `force=c(24*365, 6)`, for per="hr".
#'               Note: if `force` is not FALSE, the `overwrite` default is TRUE.
#'               DEFAULT: FALSE
#' @param overwrite Logical (vector): if force=TRUE, overwrite the existing file
#'               rather than append "_1"/"_2" etc to the filename? 
#'               DEFAULT: `!isFALSE(force)`, i.e. true when `force` is specified.
#' @param read   Logical: read the file(s) with [readDWD()]? If FALSE,
#'               only download is performed and the filename(s) returned. DEFAULT: TRUE
#' @param dbin   Logical: Download binary file, i.e. add `mode="wb"` to the
#'               [download.file()] call? 
#'               See [Website](https://bookdown.org/brry/rdwd/raster-data.html#binary-file-errors) 
#'               for details.
#'               DEFAULT: TRUE
#' @param method [download.file] `method`. Introduced in version 1.5.25 (2022-05-12)
#'               as triggered by <https://github.com/brry/rdwd/issues/34>.
#'               DEFAULT: `getOption("download.file.method")`
#' @param dfargs Named list of additional arguments passed to [download.file()]
#'               Note that mode="wb" is already passed if `dbin=TRUE`
#' @param sleep  Number. If not 0, a random number of seconds between 0 and
#'               `sleep` is passed to [Sys.sleep()] after each download
#'               to avoid getting kicked off the FTP-Server,
#'               see note in [indexFTP()]. DEFAULT: 0
#' @param progbar Logical: present a progress bar with estimated remaining time?
#'               If missing and length(file)==1, progbar is internally set to FALSE.
#'               Only works if the R package `pbapply` is available. DEFAULT: TRUE (!quiet)
#' @param browse Logical: open repository via [browseURL()] and
#'               return URL folder path? If TRUE, no data is downloaded.
#'               If file has several values, only unique folders will be opened.
#'               DEFAULT: FALSE
#' @param ntrunc Single integer: number of filenames printed in messages
#'               before they get truncated with message "(and xx more)". DEFAULT: 2
#' @param file   Deprecated since rdwd version 1.3.34, 2020-07-28.
#' @param quiet  Logical: suppress message about directory / filenames?
#'               DEFAULT: FALSE through [rdwdquiet()]
#' @param \dots  Further arguments passed to [readDWD()],
#'               like `fread`, `varnames`, `hr`, etc.
#
dataDWD <- function(
url,
base=dwdbase,
joinbf=FALSE,
dir=locdir(),
force=FALSE,
overwrite=!isFALSE(force),
read=TRUE,
dbin=TRUE,
method=getOption("download.file.method"),
dfargs=NULL,
sleep=0,
progbar=!quiet,
browse=FALSE,
ntrunc=2,
file=NULL,
quiet=rdwdquiet(),
...
)
{
if(!is.null(file)) tstop("The argument 'file' has been renamed to 'url' with rdwd version 1.3.34, 2020-07-28")
if(!is.atomic(url)) tstop("url must be a vector, not a ", class(url))
if(!is.character(url)) tstop("url must be char, not ", class(url))
base <- sub("/$","",base) # remove accidental trailing slash
url <- sub("^/","",url) # remove accidental leading slash
if(joinbf)  url <- paste0(base,"/",url)
if(missing(progbar) & length(url)==1) progbar <- FALSE
if(any(url==""))
{
  tmessage("Removing ", sum(url==""), " empty element(s) from url vector.")
  url <- url[url!=""]
}
if(length(url)<1) tstop("The vector of urls to be downloaded is empty.")
# be safe from accidental vector input:
dir     <- dir[1]
progbar <- progbar[1]
sleep   <- sleep[1]
quiet   <- quiet[1]
read    <- read[1]
browse  <- browse[1]
#
# open URL path(s) in internet browser:
if(browse)
  {
  folders <- unique(dirname(url))
  sapply(folders, browseURL)
  return(folders)
  }
# create directory to store downloaded data
owd <- dirDWD(dir, quiet=quiet)
on.exit(setwd(owd))
# output file name(s)
outfile <- gsub(paste0(base,"/"), "", url)
outfile <- gsub("/", "_", outfile)

# force=NA management
if(is.null(force)) tstop("'force' cannot be NULL. Must be TRUE, FALSE, NA or a number.")
force <- rep(force, length=length(outfile)) # recycle vector
fT <- sapply(force, isTRUE)
fF <- sapply(force, isFALSE)
if(any(fT)) force[fT] <- 0
if(any(fF)) force[fF] <- Inf
force[is.na(force)] <- 24
force <- difftime(Sys.time(), file.mtime(outfile), units="h") > force

dontdownload <- file.exists(outfile) & !force
if( any(dontdownload) & !quiet )
  {
  tmessage(sum(dontdownload), " file", if(sum(dontdownload)>1)"s",
          " already existing and not downloaded again: ",
          berryFunctions::truncMessage(outfile[dontdownload], ntrunc=ntrunc, prefix=""),
          "\nNow downloading ",sum(!dontdownload)," files...")
  }
outfile <- newFilename(outfile, quiet=quiet, ignore=dontdownload,
                       overwrite=overwrite, ntrunc=ntrunc, tellignore=FALSE)
# since berryFunctions 1.15.9 (2017-06-14), outfile is now an absolute path
# Optional progress bar:
if(progbar) lapply <- pbapply::pblapply
# ------------------------------------------------------------------------------
# loop over each filename
dl_results <- lapply(seq_along(url), function(i)
  if(!dontdownload[i])
  {
  # Actual file download:
  dfdefaults <- list(url=url[i], destfile=outfile[i], method=method, quiet=TRUE)
  if(dbin) dfdefaults <- c(dfdefaults, mode="wb")
  e <- try(suppressWarnings(do.call(download.file,
                         berryFunctions::owa(dfdefaults, dfargs))), silent=TRUE)
  # wait some time to avoid FTP bot recognition:
  if(sleep!=0) Sys.sleep(runif(n=1, min=0, max=sleep))
  return(e)
  })

# check for download errors:
iserror <- sapply(dl_results, inherits, "try-error")
if(any(iserror))
  {
  ne <- sum(iserror)
  msg <- paste0(ne, " Download", if(ne>1) "s have" else " has",
                " failed (out of ",length(iserror),").",
                if(read)" Setting read=FALSE.")
  read <- FALSE
  msg <- paste0(msg, " download.file error",if(ne>1) "s",":\n")
  msg2 <- sapply(dl_results[iserror], function(e)attr(e,"condition")$message)
  msg2 <- berryFunctions::truncMessage(msg2, ntrunc=15, prefix="", midfix="", altnix="", sep="\n")
  if(any(substr(url[iserror], 1, 4) != "ftp:"))
     msg2 <- paste0(msg2, "\n- dataDWD needs urls starting with 'ftp://'. You can use joinbf=TRUE for relative links.")
  if(grepl("cannot open URL", msg2) || grepl("Kann URL .* nicht", msg2))
     msg2 <- paste0(msg2, "\n- If files have been renamed on the DWD server, ",
                    "see   https://bookdown.org/brry/rdwd/fileindex.html")
  msg <- paste0(msg, msg2)
  warning(msg, call.=FALSE)
  }
# ------------------------------------------------------------------------------
# Output: Read the file or outfile name:
output <- outfile
if(read) output <- readDWD(file=outfile, quiet=quiet, progbar=progbar, ...)
# output:
return(invisible(output))
}
